src/sentry/snuba/outcomes.py

```diff
@@ -0,0 +1,341 @@
+            this would be used to get the number of outcomes
+
+# Tags / Grouping
+
+The Outcomes data can be grouped / filtered
+by these fields
+- `project`
+- `outcome`
+- `reason`
+- `category`
+"""
+
+ResultSet = List[Dict[str, Any]]
+Condition = Tuple[str, str, List[Any]]
+
+
+class Field(ABC):
+    @abstractmethod
+    def get_snuba_columns(self, raw_groupby: Optional[Sequence[str]] = None) -> List[str]:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def extract_from_row(
+        self, row: Optional[Mapping[str, Any]], group: Optional[Mapping[str, Any]] = None
+    ) -> int:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def aggregation(self, dataset: Dataset) -> List[str]:
+        raise NotImplementedError()
+
+
+class QuantityField(Field):
+    def get_snuba_columns(self, raw_groupby: Optional[Sequence[str]] = None) -> List[str]:
+        return ["quantity"]
+
+    def extract_from_row(
+        self, row: Optional[Mapping[str, Any]], group: Optional[Mapping[str, Any]] = None
+    ) -> int:
+        if row is None:
+            return 0
+        return int(row["quantity"])
+
+    def aggregation(self, dataset: Dataset) -> List[str]:
+        return ["sum", "quantity", "quantity"]
+
+
+class TimesSeenField(Field):
+    def get_snuba_columns(self, raw_groupby: Optional[Sequence[str]] = None) -> List[str]:
+        return ["times_seen"]
+
+    def extract_from_row(
+        self, row: Optional[Mapping[str, Any]], group: Optional[Mapping[str, Any]] = None
+    ) -> int:
+        if row is None:
+            return 0
+        return int(row["times_seen"])
+
+    def aggregation(self, dataset: Dataset) -> List[str]:
+        if dataset == Dataset.Outcomes:
+            return ["sum", "times_seen", "times_seen"]
+        else:
+            # RawOutcomes doesnt have times_seen, do a count instead
+            return ["count()", "", "times_seen"]
+
+
+class Dimension(SimpleGroupBy, ABC):
+    @abstractmethod
+    def resolve_filter(self, raw_filter: Sequence[str]) -> List[DataCategory]:
+        """
+        Based on the input filter, map it back to the clickhouse representation
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def map_row(self, row: MutableMapping[str, Any]) -> None:
+        """
+        map clickhouse representation back to presentation
+        """
+        raise NotImplementedError()
+
+
+class CategoryDimension(Dimension):
+    def resolve_filter(self, raw_filter: Sequence[str]) -> List[DataCategory]:
+        resolved_categories = set()
+        for category in raw_filter:
+            # combine DEFAULT, ERROR, and SECURITY as errors.
+            # see relay: py/sentry_relay/consts.py and relay-cabi/include/relay.h
+            if DataCategory.parse(category) == DataCategory.ERROR:
+                resolved_categories.update(DataCategory.error_categories())
+            else:
+                resolved_categories.add(DataCategory.parse(category))
```

Comment:
I think I've seen this pattern around elsewhere. Any chance a `DataCategory` class method like the following could be used from more than one place?
```python
def parse_categories(cls, category):
    dc = cls.parse(category)
    return cls.error_categories() if dc == cls.ERROR else [dc]
```

Reply:
agreed it could be useful elsewhere, but yes also agree out of scope of this changeset.